text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+
+`int<N>` and `uint<N>` will have the smallest power-of-two size and alignment.
+For example, this means that `uint<48>` will take up 8 bytes and have an
+alignment of 8, even though it only has 6 bytes of data.
+
+`int<N>` store values between -2<sup>N-1</sup> and 2<sup>N-1</sup>-1, and
+`uint<N>` stores values between 0 and 2<sup>N</sup>-1. One unexpected case of
+this is that `i1` represents zero or *negative* one, even though LLVM and other
+places use `i1` to refer to `u1`. This case is left as-is because generic code
+*should* expect sign extension to happen for all signed integers, and the
+`N = 1` case is no different. It's valid to cast `int<N>` or `uint<N>` to
+`int<M>` or `uint<M>` via `as`, and sign extension will occur as expected. This
+conversion is lossless when converting `int<N>` to `int<M>` or `uint<M>` or
+from `uint<N>` to `int<M>` or `uint<M + 1>`, where `M >= N`.
+
+In addition to the usual casts, `u1` and `i1` can also be cast *to* `bool` via
+`as`, whereas most integer types can only be cast from `bool`.
+
+For the moment, a monomorphisation error will occur if `N > 128`, to minimise
+implementation burden. This error can be lifted in the future, and if any future
+`uN` or `iN` types are added, these should alias with existing `int<N>` and
+`uint<N>` types. The main justification for this change is to allow the compiler
+to continue to represent literals and other values using `u128` internally, and
+to allow stabilising this feature *before* larger ints are allowed internally.
+
+`uint<0>` and `int<0>` are allowed as constant zero types. While this is not
+inconsistent for `uint<0>`, as zero fits in the range [0, 2<sup>n</sup>-1], it
+*is* somewhat inconsistent for `int<0>`, as the expected range for `int<0>` is
+-½ to ½. It is decided to make both of these a constant zero type but *not*
+identical types, as the invariant
+`<uint<N>>::default() == <int<N>>::default() == 0` is desired.
+
+Integer literals can be automatically coerced to `uint<N>` and `int<N>`,
+although generic `iN` and `uN` suffixes are left out for future RFCs. Because of
+the restriction that `N <= 128`, compiler code can continue to store literals as
+`u128`. When coercing literals to `uint<N>` or `int<N>`, bounds checks should be
+done just like they are with regular integer literals.
+
+Primitive operations on `int<N>` and `uint<N>` should work exactly like they do
+on `int<N>` and `uint<N>`: overflows should panic when debug assertions are
+enabled, but ignored when they are not. In general, `uint<N>` will be
+zero-extended to the next power of two, and `int<N>` will be sign-extended to
+the next power of two.
+
+## Computations and storage
+
+Because sign extension will always be applied, it's safe for the compiler to
+internally treat `uint<N>` as `uint<N.next_power_of_two()>` when doing all
+computations. As a concrete example, this means that adding two `uint<48>`
+values will work exactly like adding two `u64` values, generating exactly the
```

Comment:
This needs more thorough discussion. To take this example, an u48 add can overflow the 48 bits, setting some high bits in the 64 bit register. If we take that as given, u48 *comparisons* (to give just one example) need to first zero-extend the operands to guarantee the comparison works correctly. Conversely, we could zero-extend after arithmetic operations to get the invariant that the high 16 bits are always zero, and then use that knowledge to implement 48 bit comparisons as a plain 64 bit comparisons. Likewise for i48: you'll need sign extensions. In some code sequences compiler optimizations can prove the zero/sign extension redundant, but generally there is no free lunch here -- you need some extending even in code that never changes bit widths. Eliminating sign extensions is in fact the main reason why C compilers care about signed integer addition being UB rather than wrapping.

Reply:
I hadn't actually added this section in the original RFC when I requested feedback, and rushed it in because I felt it was necessary. You are right that in most cases, this wouldn't be a no-op, although I'm curious if optimisations could make them so.

This is certainly a case for adding `add_unchecked` and co. as suggested by… some other RFC issue I don't have the time to look up right now.